src/sema/symtable.rs

```diff
@@ -187,8 +187,19 @@ impl Symtable {
         self.scopes.push(VarScope(HashMap::new(), None));
     }
 
-    pub fn leave_scope(&mut self) {
-        self.scopes.pop();
+    pub fn leave_scope(&mut self, ns: &mut Namespace, loc: pt::Loc) {
+        if let Some(curr_scope) = self.scopes.pop() {
+            let curr_scope = curr_scope
+                .0
+                .values()
+                .filter_map(|pos| {
+                    self.vars
+                        .get(pos)
+                        .map(|var| (var.id.name.clone(), var.ty.clone()))
+                })
+                .collect();
+            ns.scopes.push((loc, curr_scope));
```

Comment:
You are storing a scope with the name of all the local variables and their types, along with the location in the global namespace. This is not good, this is only useful to the language server.

Function scopes should be stored in `struct Function` (or `struct Symtable`). Please refer to variables by their id, not their string name. The string name is not unique, one scope can shadow another variable, so the ast needs precise information - the variable id. Using the string name + type is duplicating information, waste of memory and not precise.

Reply:
I was thinking of introducing a new field in `struct Symtable` for storing the popped scopes. 
This way I can extract the variable name, type and location in the `languageserver` module and we can do away with the new `scopes` field in `struct Namespace`.